pandas/core/arrays/integer.py

```diff
@@ -0,0 +1,592 @@
+    dtype : integer dtype
+    mask : boolean 1D array, optional
+    copy : boolean, default False
+        if True, copy the input
+
+    Returns
+    -------
+    tuple of (values, mask)
+    """
+    if dtype is not None:
+        if not issubclass(type(dtype), _IntegerDtype):
+            try:
+                dtype = _dtypes[str(np.dtype(dtype))]
+            except KeyError:
+                raise ValueError("invalid dtype specified {}".format(dtype))
+
+    if isinstance(values, IntegerArray):
+        values, mask = values._data, values._mask
+        if dtype is not None:
+            values = values.astype(dtype.numpy_dtype, copy=False)
+
+        if copy:
+            values = values.copy()
+            mask = mask.copy()
+        return values, mask
+
+    values = np.array(values, copy=copy)
+    if is_object_dtype(values):
+        inferred_type = infer_dtype(values)
+        if inferred_type not in ['floating', 'integer',
+                                 'mixed-integer', 'mixed-integer-float']:
+            raise TypeError("{} cannot be converted to an IntegerDtype".format(
+                values.dtype))
+
+    elif not (is_integer_dtype(values) or is_float_dtype(values)):
+        raise TypeError("{} cannot be converted to an IntegerDtype".format(
+            values.dtype))
+
+    if mask is None:
+        mask = isna(values)
+    else:
+        assert len(mask) == len(values)
+
+    if not values.ndim == 1:
+        raise TypeError("values must be a 1D list-like")
+    if not mask.ndim == 1:
+        raise TypeError("mask must be a 1D list-like")
+
+    # infer dtype if needed
+    if dtype is None:
+        if is_integer_dtype(values):
+            dtype = values.dtype
+        else:
+            dtype = np.dtype('int64')
+    else:
+        dtype = dtype.type
+
+    # if we are float, let's make sure that we can
+    # safely cast
+
+    # we copy as need to coerce here
+    if mask.any():
+        values = values.copy()
+        values[mask] = 1
+        values = safe_cast(values, dtype, copy=False)
+    else:
+        values = safe_cast(values, dtype, copy=False)
+
+    return values, mask
+
+
+class IntegerArray(ExtensionArray, ExtensionOpsMixin):
+    """
+    We represent an IntegerArray with 2 numpy arrays
+    - data: contains a numpy integer array of the appropriate dtype
+    - mask: a boolean array holding a mask on the data, False is missing
+    """
+
+    @cache_readonly
+    def dtype(self):
+        return _dtypes[str(self._data.dtype)]
+
+    def __init__(self, values, mask=None, dtype=None, copy=False):
+        """
+        Parameters
+        ----------
+        values : 1D list-like / IntegerArray
+        mask : 1D list-like, optional
+        dtype : subclass of _IntegerDtype, optional
+        copy : bool, default False
+
+        Returns
+        -------
+        IntegerArray
+        """
+        self._data, self._mask = coerce_to_array(
+            values, dtype=dtype, mask=mask, copy=copy)
```

Comment:
this is not possible, as quite a lot of checking / coercion needs to occur on the input. e.g. make it an ndarray if not, coerce dtype. this is the point of ``coerce_to_array``, to make it DRY (as I also have to do this with setitem).

Reply:
It *is* possible, just a different API design choice. 
I personally think it makes it cleaner code-wise, to clearly separate those case where coercion and checking is needed (there we would still use `coerce_to_array` or another method calling that under the hood) and those internal places where it is not needed (as those also exist: eg getitem). 
It is very similar to the `fastpath` keyword in many constructors, but I think we agree on is not a good design to mix both in the same constructor.

Reply:
I agree with @jorisvandenbossche, we certainly could make a simpler constructor which doesn't call `coerce_to_array()` internally. This is a perfect use case for an alternative constructor, e.g., `IntegerArray.from_values()`.

In general I think this is a better/cleaner pattern to use for internal APIs in pandas, because it involves less introspection (which tends to be error/bug prone).

Reply:
did you see my response above?

Reply:
`pd.Series(array, dtype='Int64')` could call `IntegerArray.from_values(array, dtype='Int64')` internally. It doesn't need to call the `IntegerArray()` constructor directly.

Reply:
Can you respond here further?